doc/contributing.rst

```diff
@@ -1,221 +1,577 @@
 
-Importing
-^^^^^^^^^
-* Import modules in this order:
-  1. builtin
-  2. standard scientific (``numpy as np``, ``scipy`` submodules)
-  3. others
-  4. mne imports (relative within the MNE module, absolute in the examples)
-* Imports for ``matplotlib`` and optional modules (``sklearn``, and ``pandas``, etc.) within the MNE module should be nested (i.e., within a function or method, not at the top of a file).
+Next, we'll do the same thing for Mayavi (one of our 3D plotting backends):
 
-Vizualization
-^^^^^^^^^^^^^
-* Add public functions to the :mod:`mne.viz` package and use these in the corresponding methods.
-* All visualization functions must accept a ``show`` parameter and return a ``fig`` handle.
-* Use ``RdBu_r`` colormap for signed data with a meaningful middle (zero-point) and ``Reds`` otherwise in visualization functions and examples.
+.. code-block:: console
 
-Return types
-^^^^^^^^^^^^
-* Methods should modify inplace and return ``self``, functions should return copies (where applicable).
-
-Style
-^^^^^
-* Use single quotes whenever possible.
-* Prefer generator or list comprehensions over ``filter``, ``map`` and other functional idioms.
-* Use explicit functional constructors for builtin containers to improve readability (e.g., ``list()``, ``dict``).
-* Avoid nested functions or class methods if possible -- use private functions instead.
-* Avoid ``**kwargs`` and ``*args`` in function signatures.
-* Add brief docstrings to simple private functions and complete docstrings for complex ones.
-
-Checking and building documentation
------------------------------------
-
-All changes to the codebase must be properly documented.
-To ensure that documentation is rendered correctly, the best bet is to
-follow the existing examples for class and function docstrings,
-and examples and tutorials.
-
-Our documentation (including docstring in code) uses ReStructuredText format,
-see `Sphinx documentation`_ to learn more about editing them. Our code
-follows the `NumPy docstring standard`_.
-
-Documentation is automatically built remotely during pull requests. If
-you want to also test documentation locally, you will need to install (using
-``conda`` or ``pip`` as preferred):
+    $ pip uninstall -y mayavi
+    $ cd $INSTALL_LOCATION
+    $ git clone git://github.com/enthought/mayavi.git
+    $ cd mayavi
+    $ pip install -e .  # or: python setup.py develop
+
+Finally, add a few dependencies that are not needed for running MNE-Python but
+are needed for locally running our test suite or building our documentation:
+
+.. code-block:: console
+
+    $ pip install sphinx sphinx-gallery sphinx_bootstrap_theme sphinx_fontawesome memory_profiler
+    $ conda install sphinx-autobuild doc8  # linter for reStructuredText
+
+.. TODO: no longer needed? pip install sphinx-gallery should be fine...
+    $ cd $INSTALL_LOCATION
+    $ git clone git://github.com/sphinx-gallery/sphinx-gallery.git
+    $ cd sphinx-gallery
+    $ pip install -e .
```

Comment:
I think developers should know how to install the development version for _most_ package. In case a bug is found, they should have the know how to switch between the stable and development version.

Reply:
Furthermore, there is a one liner pip command that will give you the same thing except that it won't be editable:

```sh
$ pip install https://api.github.com/repos/sphinx-gallery/sphinx-gallery/zipball/master
```

Reply:
Maybe we should add a `.. note:` after the installation of MNE dev version, reminding folks that if they need a very recent bug fix for *any* module, they can do one-time pip+git updates (and point then to the `advanced_setup` page where we demo that already), or they can do what we did here (a full clone and `pip -e`). WDYT?

Reply:
yes, pointing to an advanced setup page sounds like a good idea to me!

Reply:
I am always using `conda develop .` from conda-build. Is there any reason not to use this put `pip install -e` instead? Or vice versa? Maybe a short note as to those alternatives would be nice.

Reply:
the reason we don't mention `conda develop .` is that it requires users to install `conda-build` which they otherwise wouldn't need for installation/setup. I am -1 for adding a note about it, since the contributor guide is already very long, and users sophisticated enough to be running `conda-build` will probably already know enough to recognize that they can substitute `conda develop .` here (without needing us to tell them that it would work).

Reply:
Agreed let's just put it as a standard `pip` install to KISS. It can be the `master` one if we want. MNE maintainers are pretty active on SG so it's not a bad idea to keep up to date there as it often has some bugfixes.